ssh-key/src/sshsig.rs

```diff
@@ -0,0 +1,328 @@
+//! `sshsig` implementation.
+
+use crate::{
+    checked::CheckedSum,
+    decode::Decode,
+    encode::Encode,
+    pem::{self, PemLabel},
+    public,
+    reader::Reader,
+    writer::Writer,
+    Algorithm, Error, HashAlg, Result, Signature, SigningKey, PEM_LINE_WIDTH,
+};
+use alloc::{borrow::ToOwned, string::String, string::ToString, vec::Vec};
+use base64ct::LineEnding;
+use core::str::FromStr;
+use signature::{Signer, Verifier};
+
+type Version = u32;
+
+/// `sshsig` provides a general-purpose signature format based on SSH keys and
+/// wire formats.
+///
+/// These signatures can be produced using `ssh-keygen -Y sign`. They're
+/// encoded as PEM and begin with the following:
+///
+/// ```text
+/// -----BEGIN SSH SIGNATURE-----
+/// ```
+///
+/// See [PROTOCOL.sshsig] for more information.
+///
+/// [PROTOCOL.sshsig]: https://cvsweb.openbsd.org/src/usr.bin/ssh/PROTOCOL.sshsig?annotate=HEAD
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct SshSig {
+    version: Version,
+    public_key: public::KeyData,
+    namespace: String,
+    reserved: Vec<u8>,
+    hash_alg: HashAlg,
+    signature: Signature,
+}
+
+impl SshSig {
+    /// Supported version.
+    pub const VERSION: Version = 1;
+
+    /// The preamble is the six-byte sequence "SSHSIG".
+    ///
+    /// It is included to ensure that manual signatures can never be confused
+    /// with any message signed during SSH user or host authentication.
+    const MAGIC_PREAMBLE: &'static [u8] = b"SSHSIG";
+
+    /// Decode signature from PEM which begins with the following:
+    ///
+    /// ```text
+    /// -----BEGIN SSH SIGNATURE-----
+    /// ```
+    pub fn from_pem(pem: impl AsRef<[u8]>) -> Result<Self> {
+        let mut reader = pem::Decoder::new_wrapped(pem.as_ref(), PEM_LINE_WIDTH)?;
+        Self::validate_pem_label(reader.type_label())?;
+        let signature = Self::decode(&mut reader)?;
+        reader.finish(signature)
+    }
+
+    /// Encode signature as PEM which begins with the following:
+    ///
+    /// ```text
+    /// -----BEGIN SSH SIGNATURE-----
+    /// ```
+    pub fn to_pem(&self, line_ending: LineEnding) -> Result<String> {
+        let encoded_len = pem::encapsulated_len_wrapped(
+            Self::PEM_LABEL,
+            PEM_LINE_WIDTH,
+            line_ending,
+            self.encoded_len()?,
+        )?;
+
+        let mut buf = vec![0u8; encoded_len];
+        let mut writer =
+            pem::Encoder::new_wrapped(Self::PEM_LABEL, PEM_LINE_WIDTH, line_ending, &mut buf)?;
+
+        self.encode(&mut writer)?;
+        let actual_len = writer.finish()?;
+        buf.truncate(actual_len);
+        Ok(String::from_utf8(buf)?)
+    }
+
+    /// Sign the given message with the provided signing key.
+    pub fn sign<S: SigningKey>(
```

Comment:
This being the only way to construct a `SignedData` is a little awkward for what I’m doing: I’m using an agent to obtain signatures, yet I don’t want to pull in an async runtime just for that. To make it satisfy `SigningKey`, I’d need to slap a mutex around the unix socket, while a `SignerMut` would otherwise do.

Could `SignedData` be made public, perhaps without revealing its structure? Or alternatively, a second `sign` method which allows a mutable signer?

Reply:
Instead of exposing the `SignedData` struct, there could be a public method to get the serialized signed data as a `Vec<u8>`, and a public constructor for `SshSig` that accepts a `Signature` over it.

Reply:
Yeah that would work for me.

Reply:
Added in #33

Reply:
Thanks!